\documentclass[9pt,english]{extarticle}\usepackage[]{graphicx}\usepackage[]{color}
%% maxwidth is the original width if it is less than linewidth
%% otherwise use linewidth (to make sure the graphics do not exceed the margin)
\makeatletter
\def\maxwidth{ %
  \ifdim\Gin@nat@width>\linewidth
    \linewidth
  \else
    \Gin@nat@width
  \fi
}
\makeatother

\definecolor{fgcolor}{rgb}{0.345, 0.345, 0.345}
\newcommand{\hlnum}[1]{\textcolor[rgb]{0.686,0.059,0.569}{#1}}%
\newcommand{\hlstr}[1]{\textcolor[rgb]{0.192,0.494,0.8}{#1}}%
\newcommand{\hlcom}[1]{\textcolor[rgb]{0.678,0.584,0.686}{\textit{#1}}}%
\newcommand{\hlopt}[1]{\textcolor[rgb]{0,0,0}{#1}}%
\newcommand{\hlstd}[1]{\textcolor[rgb]{0.345,0.345,0.345}{#1}}%
\newcommand{\hlkwa}[1]{\textcolor[rgb]{0.161,0.373,0.58}{\textbf{#1}}}%
\newcommand{\hlkwb}[1]{\textcolor[rgb]{0.69,0.353,0.396}{#1}}%
\newcommand{\hlkwc}[1]{\textcolor[rgb]{0.333,0.667,0.333}{#1}}%
\newcommand{\hlkwd}[1]{\textcolor[rgb]{0.737,0.353,0.396}{\textbf{#1}}}%

\usepackage{framed}
\makeatletter
\newenvironment{kframe}{%
 \def\at@end@of@kframe{}%
 \ifinner\ifhmode%
  \def\at@end@of@kframe{\end{minipage}}%
  \begin{minipage}{\columnwidth}%
 \fi\fi%
 \def\FrameCommand##1{\hskip\@totalleftmargin \hskip-\fboxsep
 \colorbox{shadecolor}{##1}\hskip-\fboxsep
     % There is no \\@totalrightmargin, so:
     \hskip-\linewidth \hskip-\@totalleftmargin \hskip\columnwidth}%
 \MakeFramed {\advance\hsize-\width
   \@totalleftmargin\z@ \linewidth\hsize
   \@setminipage}}%
 {\par\unskip\endMakeFramed%
 \at@end@of@kframe}
\makeatother

\definecolor{shadecolor}{rgb}{.97, .97, .97}
\definecolor{messagecolor}{rgb}{0, 0, 0}
\definecolor{warningcolor}{rgb}{1, 0, 1}
\definecolor{errorcolor}{rgb}{1, 0, 0}
\newenvironment{knitrout}{}{} % an empty environment to be redefined in TeX

\usepackage{alltt}
\include{canevas_config}
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\begin{document}
\author{Sleiman Bassim, PhD}
\title{R implementation}
\maketitle
\begin{linenumbers}



Locally loaded functions:
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlkwd{source}\hlstd{(}\hlstr{"~/data/Dropbox/humanR/01funcs.R"}\hlstd{)}
\hlcom{#load("", .GlobalEnv)}
\hlkwd{lsos}\hlstd{(}\hlkwc{pat}\hlstd{=}\hlstr{""}\hlstd{)}
\end{alltt}
\begin{verbatim}
                         Type    Size PrettySize  Rows Columns
a.df                   tbl_df    6848     6.7 Kb    42       2
a.list                   list    6144       6 Kb     2      NA
bagging              function   23520      23 Kb    NA      NA
bagging.clas         function   24280    23.7 Kb    NA      NA
baggingTune          function   23856    23.3 Kb    NA      NA
circos.test          function   55408    54.1 Kb    NA      NA
Cols                 function    6576     6.4 Kb    NA      NA
ensemble.mean        function   24232    23.7 Kb    NA      NA
i                     integer      48   48 bytes     1      NA
indices.df            integer    4040     3.9 Kb  1000      NA
input.df           data.frame 2432000     2.3 Mb 10027      20
input.subset.df        tbl_df 1497248     1.4 Mb  1000      20
input.summary        function   12808    12.5 Kb    NA      NA
locusRMR             function   11224      11 Kb    NA      NA
model.clas           function   20680    20.2 Kb    NA      NA
model.reg            function   18528    18.1 Kb    NA      NA
modelTune.clas       function   21016    20.5 Kb    NA      NA
modelTune.reg        function   18864    18.4 Kb    NA      NA
net2Bin              function   33240    32.5 Kb    NA      NA
normDataWithin       function   21816    21.3 Kb    NA      NA
olBarplot            function   31992    31.2 Kb    NA      NA
output.df          data.frame    3568     3.5 Kb    46       3
output.df.log      data.frame    7304     7.1 Kb    29       3
output.df.reshape     cast_df   15944    15.6 Kb    21       3
output.df.stand        matrix     640  640 bytes    29       1
output.df.variants data.frame    7208       7 Kb    29       3
overLapper           function   86584    84.6 Kb    NA      NA
pkgs                character     392  392 bytes     6      NA
predict.regsubsets   function   10208      10 Kb    NA      NA
range_correlation    function   16472    16.1 Kb    NA      NA
regfit               function    9912     9.7 Kb    NA      NA
regfit.int           function   10360    10.1 Kb    NA      NA
rocplot              function    7768     7.6 Kb    NA      NA
set.cor              function   10712    10.5 Kb    NA      NA
set.var              function   12216    11.9 Kb    NA      NA
summarySE            function   30400    29.7 Kb    NA      NA
summarySEwithin      function   39616    38.7 Kb    NA      NA
testing                  list 2882440     2.7 Mb     2      NA
top100                 tbl_df   70032    68.4 Kb   103      12
vennPlot             function  562632   549.4 Kb    NA      NA
visualizeNet         function   30288    29.6 Kb    NA      NA
\end{verbatim}
\end{kframe}
\end{knitrout}

\section{By-group aggregation}
\label{sec:aggregation}

By-group aggregation uses the principle of split-and-conquer or \textbf{split-apply-combine} a term coined by \href{http://goo.gl/qZIQ8I}{Hadley Wickham}. R jobs will run in parallel across several cores which reduces the calculation time of the row search. 


\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlstd{pkgs} \hlkwb{<-} \hlkwd{c}\hlstd{(}\hlstr{'rbenchmark'}\hlstd{,} \hlstr{'doParallel'}\hlstd{,} \hlstr{'foreach'}\hlstd{,} \hlstr{'Hmisc'}\hlstd{,} \hlstr{'dplyr'}\hlstd{,} \hlstr{'plyr'}\hlstd{)}
\hlkwd{lapply}\hlstd{(pkgs, require,} \hlkwc{character.only} \hlstd{=} \hlnum{TRUE}\hlstd{);}
\end{alltt}
\begin{verbatim}
[[1]]
[1] TRUE

[[2]]
[1] TRUE

[[3]]
[1] TRUE

[[4]]
[1] TRUE

[[5]]
[1] TRUE

[[6]]
[1] TRUE
\end{verbatim}
\end{kframe}
\end{knitrout}

Get data from this site \href{http://goo.gl/nB0gAH}{here}. It is in *.xlsx form. Use \verb|read.xlsx| from the package \verb|xlsx| to read in the table into an R dataframe. Put it in a wrapped dataframe with \verb|dplyr| and saved it in a .Rdata file.
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlkwd{load}\hlstd{(}\hlstr{"LoF_Phase1_v3.Rdata"}\hlstd{, .GlobalEnv)}
\hlkwd{lsos}\hlstd{(}\hlkwc{pattern}\hlstd{=}\hlstr{"input.*"}\hlstd{)}
\end{alltt}
\begin{verbatim}
                      Type    Size PrettySize  Rows Columns
input.df        data.frame 2432000     2.3 Mb 10027      20
input.subset.df     tbl_df 1497248     1.4 Mb  1000      20
input.summary     function   12808    12.5 Kb    NA      NA
\end{verbatim}
\end{kframe}
\end{knitrout}


Sampling 1000 rows from the dataset is followed with a grouping and summarizing functions to show detail of the data and the levels of classification.
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlstd{indices.df} \hlkwb{<-} \hlkwd{sample}\hlstd{(}\hlnum{1}\hlopt{:}\hlkwd{nrow}\hlstd{(input.df),}\hlnum{1000}\hlstd{)}
\hlcom{#input.df <- input.df[complete.cases(input.df), ]}
\hlstd{input.subset.df} \hlkwb{<-} \hlkwd{tbl_df}\hlstd{(input.df[indices.df, ])}
\hlkwd{str}\hlstd{(input.subset.df)}
\end{alltt}
\begin{verbatim}
Classes 'tbl_df', 'tbl' and 'data.frame':	1000 obs. of  20 variables:
 $ chr             : Factor w/ 23 levels "1","10","11",..: 2 8 11 13 21 12 17 1 16 1 ...
 $ pos             : num  101180550 48177825 53116811 20033367 145139945 ...
 $ rsID            : Factor w/ 5383 levels ".","rs10009430",..: 1541 1 2426 5056 3447 2223 1 4182 4576 1 ...
 $ ref_allele      : Factor w/ 145 levels "A","AAACC","AAC",..: 107 107 33 33 33 107 107 1 75 68 ...
 $ alt_allele      : Factor w/ 89 levels "A","AAC","AAG",..: 20 1 65 40 65 1 1 40 65 20 ...
 $ ancestral_allele: Factor w/ 125 levels "-",".","A","a",..: 94 94 93 29 29 94 94 3 67 61 ...
 $ effect          : Factor w/ 2 levels "full","partial": 2 1 1 2 1 1 2 2 1 1 ...
 $ type            : Factor w/ 3 levels "frameshift_indel",..: 2 3 3 2 2 3 2 2 3 1 ...
 $ an              : num  2184 2184 2184 2184 2184 ...
 $ ac              : num  1 1 49 561 1 2 1 1 1 67 ...
 $ gene            : Factor w/ 6433 levels "A1BG","A2M","A2ML1",..: 2597 33 6404 1698 2603 4750 797 1013 1088 1900 ...
 $ gene_id         : Factor w/ 6433 levels "ENSG00000000419.7",..: 1616 2596 3959 860 5563 213 5310 2678 5062 3623 ...
 $ lof_trans       : num  1 4 7 1 2 3 2 1 1 3 ...
 $ all_trans       : num  3 4 7 3 2 3 3 3 1 3 ...
 $ failed_filters  : num  1 0 0 1 1 0 0 0 0 0 ...
 $ filters_failed  : Factor w/ 69 levels "0","all_alt",..: 46 1 1 65 46 1 1 1 1 1 ...
 $ splice_type     : Factor w/ 3 levels ".","acceptor",..: 3 1 1 2 2 1 2 3 1 1 ...
 $ canonical       : Factor w/ 3 levels ".","NO","YES": 2 1 1 3 2 1 3 3 1 1 ...
 $ other_canonical : Factor w/ 3 levels ".","NO","YES": 2 1 1 3 2 1 3 3 1 1 ...
 $ intron_length   : Factor w/ 2266 levels ".","1","100",..: 344 1 1 1177 523 1 2002 2006 1 1 ...
\end{verbatim}
\begin{alltt}
\hlkwd{unique}\hlstd{(input.subset.df}\hlopt{$}\hlstd{effect)}
\end{alltt}
\begin{verbatim}
[1] partial full    <NA>   
Levels: full partial
\end{verbatim}
\begin{alltt}
\hlkwd{unique}\hlstd{(input.subset.df}\hlopt{$}\hlstd{chr)}
\end{alltt}
\begin{verbatim}
 [1] 10   16   19   20   8    2    4    1    3    22   5    14   12   6    9    11  
[17] 7    15   18   13   21   17   X    <NA>
Levels: 1 10 11 12 13 14 15 16 17 18 19 2 20 21 22 3 4 5 6 7 8 9 X
\end{verbatim}
\end{kframe}
\end{knitrout}

Extract subsets of the dataframe depending on the "effect" column.
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlstd{input.summary} \hlkwb{<-} \hlkwa{function}\hlstd{(}\hlkwc{x}\hlstd{)\{}
  \hlstd{y} \hlkwb{<-} \hlkwd{unique}\hlstd{(x)}
  \hlstd{sub} \hlkwb{=} \hlkwd{list}\hlstd{(}\hlkwa{NULL}\hlstd{)}
  \hlkwa{for}\hlstd{(i} \hlkwa{in} \hlnum{1}\hlopt{:}\hlkwd{length}\hlstd{(y))\{}
    \hlstd{sub[[i]]} \hlkwb{<-} \hlkwd{subset}\hlstd{(input.subset.df, x} \hlopt{==} \hlstd{y[i])}
  \hlstd{\}}
  \hlstd{sub}
\hlstd{\}}

\hlstd{testing} \hlkwb{<-} \hlkwd{input.summary}\hlstd{(input.subset.df}\hlopt{$}\hlstd{effect)}
\hlstd{testing[}\hlnum{1}\hlstd{]}
\end{alltt}
\begin{verbatim}
[[1]]
Source: local data frame [463 x 20]

   chr       pos        rsID ref_allele alt_allele ancestral_allele  effect
1   10 101180550 rs145279669          T          C                T partial
2   20  20033367   rs7508949          C          G                C partial
3    4  42146013           .          T          A                T partial
4    1   1888057 rs191790164          A          G                A partial
5    2 241555861 rs116229176          G          A                G partial
6    2 169923686 rs151018449          C          A                C partial
7    4 170634334           .          C          T                C partial
8    8  69699676 rs150041402          A          C                A partial
9    3  46785355   rs4075012          T          C                C partial
10  19   2340154           .          G         GC               GC partial
.. ...       ...         ...        ...        ...              ...     ...
Variables not shown: type (fctr), an (dbl), ac (dbl), gene (fctr), gene_id (fctr),
  lof_trans (dbl), all_trans (dbl), failed_filters (dbl), filters_failed (fctr),
  splice_type (fctr), canonical (fctr), other_canonical (fctr), intron_length (fctr)
\end{verbatim}
\begin{alltt}
\hlstd{testing[}\hlnum{2}\hlstd{]}
\end{alltt}
\begin{verbatim}
[[1]]
Source: local data frame [536 x 20]

   chr       pos        rsID ref_allele alt_allele ancestral_allele effect
1   16  48177825           .          T          A                T   full
2   19  53116811  rs17855778          C          T                N   full
3    8 145139945 rs187027021          C          T                C   full
4    2  85581474 rs150217356          T          A                T   full
5    3  49842325  rs35389534          G          T                G   full
6    1  40981007           .        CTG          C              CTG   full
7   22  37398102           .          G          A                .   full
8    5 148680794           .          G          A                G   full
9   14  98100029 rs182090492          C          T                C   full
10  12  11461802  rs12829245          G          A                .   full
.. ...       ...         ...        ...        ...              ...    ...
Variables not shown: type (fctr), an (dbl), ac (dbl), gene (fctr), gene_id (fctr),
  lof_trans (dbl), all_trans (dbl), failed_filters (dbl), filters_failed (fctr),
  splice_type (fctr), canonical (fctr), other_canonical (fctr), intron_length (fctr)
\end{verbatim}
\end{kframe}
\end{knitrout}

\section{Different grouping and summarizing Functions}
\label{sec:functions}

There is several functions to be used to extract a specific amount of data.
The \verb|tapply, agreggate, ddply| functions:
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlkwd{tapply}\hlstd{(input.subset.df}\hlopt{$}\hlstd{pos, input.subset.df}\hlopt{$}\hlstd{effect, mean)}
\end{alltt}
\begin{verbatim}
    full  partial 
73933120 74055359 
\end{verbatim}
\end{kframe}
\end{knitrout}

Extract and visualize a summary of the partial and full effects of variants relative to the nature of the chromosome.
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlkwd{require}\hlstd{(plyr)}
\hlstd{output.df} \hlkwb{<-} \hlkwd{ddply}\hlstd{(}
    \hlstd{input.subset.df,}
    \hlkwd{c}\hlstd{(}\hlstr{"effect"}\hlstd{,}\hlstr{"chr"}\hlstd{),}
    \hlkwa{function}\hlstd{(}\hlkwc{x}\hlstd{)\{}
        \hlkwd{c}\hlstd{(}\hlkwc{count} \hlstd{=} \hlkwd{nrow}\hlstd{(x))}
    \hlstd{\})}
\hlcom{## Plot}
\hlkwd{require}\hlstd{(ggplot2)}
\hlkwd{ggplot}\hlstd{(output.df)} \hlopt{+}
    \hlkwd{geom_bar}\hlstd{(}\hlkwd{aes}\hlstd{(}\hlkwc{x} \hlstd{= chr,} \hlkwc{y} \hlstd{= count,} \hlkwc{fill} \hlstd{= effect),}
             \hlkwc{stat} \hlstd{=} \hlstr{'identity'}\hlstd{,} \hlkwc{position} \hlstd{=} \hlstr{'dodge'}\hlstd{,} \hlkwc{width} \hlstd{=} \hlnum{.7}\hlstd{)} \hlopt{+}
    \hlkwd{scale_fill_manual}\hlstd{(}\hlstr{"Variant Effect\textbackslash{}n"}\hlstd{,} \hlkwc{values} \hlstd{=} \hlkwd{c}\hlstd{(}\hlstr{"grey"}\hlstd{,} \hlstr{"black"}\hlstd{),}
                      \hlkwc{labels} \hlstd{=} \hlkwd{c}\hlstd{(}\hlstr{'full'}\hlstd{,} \hlstr{'partial'}\hlstd{))} \hlopt{+}
    \hlkwd{labs}\hlstd{(}\hlkwc{x} \hlstd{=} \hlstr{'\textbackslash{}nChromosome'}\hlstd{,} \hlstr{'Count\textbackslash{}n'}\hlstd{)} \hlopt{+}
    \hlkwd{theme_bw}\hlstd{()}
\end{alltt}
\end{kframe}

{\centering \includegraphics[width=\linewidth,height=.45\linewidth]{graphics/plotcomaprison-1} 

}



\end{knitrout}

Visualize a count summary relative to the nature of variants.
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlstd{output.df.variants} \hlkwb{<-} \hlkwd{ddply}\hlstd{(}
    \hlstd{input.subset.df,}
    \hlkwd{c}\hlstd{(}\hlstr{'effect'}\hlstd{,}\hlstr{'alt_allele'}\hlstd{),}
    \hlkwa{function}\hlstd{(}\hlkwc{x}\hlstd{)\{}
    \hlkwd{c}\hlstd{(}\hlkwc{count} \hlstd{=} \hlkwd{nrow}\hlstd{(x))}
    \hlstd{\})}
\hlkwd{head}\hlstd{(output.df.variants)}
\end{alltt}
\begin{verbatim}
  effect alt_allele count
1   full          A   225
2   full         AG     1
3   full          C    49
4   full         CA     2
5   full     CACACT     1
6   full         CG     3
\end{verbatim}
\begin{alltt}
\hlkwd{require}\hlstd{(reshape)}
\hlstd{output.df.reshape} \hlkwb{<-} \hlkwd{cast}\hlstd{(output.df.variants, alt_allele}\hlopt{~}\hlstd{effect)}
\hlkwd{head}\hlstd{(output.df.reshape)}
\end{alltt}
\begin{verbatim}
  alt_allele full partial NA
1          A  225     148 NA
2         AG    1       2 NA
3        AGG   NA       1 NA
4          C   49      53 NA
5         CA    2       1 NA
6     CACACT    1      NA NA
\end{verbatim}
\begin{alltt}
\hlstd{output.df.reshape[}\hlkwd{is.na}\hlstd{(output.df.reshape)]} \hlkwb{<-} \hlnum{0}
\end{alltt}


{\ttfamily\noindent\color{warningcolor}{Warning in `[<-.factor`(`*tmp*`, thisvar, value = 0): invalid factor level, NA generated}}\begin{alltt}
\hlkwd{require}\hlstd{(vegan)}
\hlstd{output.df.stand} \hlkwb{<-} \hlkwd{decostand}\hlstd{(output.df.variants}\hlopt{$}\hlstd{count,} \hlkwc{method}\hlstd{=}\hlstr{"log"}\hlstd{)}
\hlstd{output.df.log} \hlkwb{<-} \hlkwd{data.frame}\hlstd{(}\hlkwc{effect} \hlstd{= output.df.variants}\hlopt{$}\hlstd{effect,}
                            \hlkwc{pos}\hlstd{= output.df.variants}\hlopt{$}\hlstd{alt_allele,}
                            \hlkwc{count}\hlstd{= output.df.stand[,}\hlnum{1}\hlstd{])}

\hlkwd{ggplot}\hlstd{(output.df.log)} \hlopt{+}
    \hlkwd{geom_bar}\hlstd{(}\hlkwd{aes}\hlstd{(}\hlkwc{x} \hlstd{= pos,} \hlkwc{y} \hlstd{= count,} \hlkwc{fill} \hlstd{= effect),}
             \hlkwc{stat} \hlstd{=} \hlstr{'identity'}\hlstd{,} \hlkwc{position} \hlstd{=} \hlstr{'dodge'}\hlstd{,} \hlkwc{width} \hlstd{=} \hlnum{.7}\hlstd{)} \hlopt{+}
    \hlkwd{scale_fill_manual}\hlstd{(}\hlstr{"Variant Effect\textbackslash{}n"}\hlstd{,} \hlkwc{values} \hlstd{=} \hlkwd{c}\hlstd{(}\hlstr{"grey"}\hlstd{,} \hlstr{"black"}\hlstd{),}
                      \hlkwc{labels} \hlstd{=} \hlkwd{c}\hlstd{(}\hlstr{'full'}\hlstd{,} \hlstr{'partial'}\hlstd{))} \hlopt{+}
    \hlkwd{labs}\hlstd{(}\hlkwc{x} \hlstd{=} \hlstr{'\textbackslash{}nChromosome'}\hlstd{,} \hlstr{'Count\textbackslash{}n'}\hlstd{)} \hlopt{+}
    \hlkwd{theme_bw}\hlstd{()}
\end{alltt}
\end{kframe}

{\centering \includegraphics[width=\linewidth,height=.5\linewidth]{graphics/plotallele-1} 

}



\end{knitrout}

Create a data frame from a list.
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlstd{a.list} \hlkwb{<-} \hlkwd{list}\hlstd{(}\hlkwc{x}\hlstd{=output.df.reshape[,}\hlnum{1}\hlstd{],} \hlkwc{y}\hlstd{=output.df.reshape[,}\hlnum{2}\hlstd{])}
\hlkwd{system.time}\hlstd{(a.df} \hlkwb{<-} \hlkwd{tbl_df}\hlstd{(}\hlkwd{do.call}\hlstd{(}\hlstr{"rbind.fill"}\hlstd{,}\hlkwd{lapply}\hlstd{(a.list, as.data.frame))))}
\end{alltt}
\begin{verbatim}
   user  system elapsed 
  0.004   0.000   0.002 
\end{verbatim}
\end{kframe}
\end{knitrout}

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlkwd{registerDoParallel}\hlstd{(}\hlkwc{cores} \hlstd{=} \hlnum{2}\hlstd{)}
\hlkwd{benchmark}\hlstd{(}\hlkwc{replications}\hlstd{=}\hlnum{100}\hlstd{,} \hlkwd{lapply}\hlstd{(}\hlnum{1}\hlopt{:}\hlnum{3}\hlstd{, sqrt),} \hlkwd{foreach}\hlstd{(}\hlkwc{i}\hlstd{=}\hlnum{1}\hlopt{:}\hlnum{3}\hlstd{)} \hlopt{%do%} \hlkwd{sqrt}\hlstd{(i))}
\end{alltt}
\begin{verbatim}
                           test replications elapsed relative user.self sys.self
2 foreach(i = 1:3) %do% sqrt(i)          100   0.575      575      0.54    0.028
1             lapply(1:3, sqrt)          100   0.001        1      0.00    0.000
  user.child sys.child
2          0         0
1          0         0
\end{verbatim}
\end{kframe}
\end{knitrout}

\subsection{More dplyr verbs for data manipulation}
\label{subsec:verbs}
Verbs in \verb|dplyr| constitute the essential functions for data manipulation. Five exist and are basic dataframe manipulation functions. \textbf{Filter} to choose rows, \textbf{select} to choose from columns, \textbf{arrange} to order rows, \textbf{mutate} and \textbf{summarize} to make new columns.
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlstd{input.subset.df} \hlopt{%>%}
    \hlkwd{select}\hlstd{(chr, pos, ref_allele, effect)} \hlopt{%>%}
    \hlkwd{filter}\hlstd{(ref_allele} \hlopt{==} \hlstr{"A"}\hlstd{)} \hlopt{%>%}
    \hlkwd{group_by}\hlstd{(chr, effect)} \hlopt{%>%}
    \hlkwd{mutate}\hlstd{(}\hlkwc{pos2} \hlstd{=} \hlkwd{rank}\hlstd{(pos))} \hlopt{%>%}
    \hlkwd{filter}\hlstd{(pos} \hlopt{<} \hlnum{8.7}\hlopt{*}\hlnum{10}\hlopt{^}\hlnum{6} \hlopt{&} \hlstd{pos} \hlopt{>} \hlnum{8}\hlopt{*}\hlnum{10}\hlopt{^}\hlnum{6} \hlstd{)} \hlopt{%>%}
    \hlkwd{arrange}\hlstd{(chr, effect, pos2)}
\end{alltt}
\begin{verbatim}
Source: local data frame [2 x 5]
Groups: chr, effect

  chr     pos ref_allele  effect pos2
1   1 8585816          A partial    9
2  12 8288179          A    full    8
\end{verbatim}
\end{kframe}
\end{knitrout}

\section{Other options of parallel computing}
For MySQL try \verb|dbApply()| to aggregate. And \verb|sqldf| will create a temporary database. Or \href{www.bigmemory.org}{bigmemory.org} to load data into memory, or use Hadoop.


\section{Top 100 papers cited in Nature}
Download data from onedrive in xlsx form.
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlcom{#require(xlsx)}
\hlcom{#top100 <- read.xlsx("~/Downloads/WebofSciencetop100.xlsx", sheetIndex = "Sheet1")}
\hlkwd{load}\hlstd{(}\hlstr{"WebofScienceTop100.Rdata"}\hlstd{, .GlobalEnv)}
\hlstd{top100} \hlkwb{<-} \hlkwd{tbl_df}\hlstd{(top100)}
\hlkwd{plot}\hlstd{(top100}\hlopt{$}\hlstd{Year,top100}\hlopt{$}\hlstd{Rnk,} \hlkwc{pch} \hlstd{=} \hlnum{20}\hlstd{)}
\hlkwd{abline}\hlstd{(}\hlkwd{lm}\hlstd{(top100}\hlopt{$}\hlstd{Rank}\hlopt{~}\hlstd{top100}\hlopt{$}\hlstd{Year),} \hlkwc{col} \hlstd{=} \hlstr{"red"}\hlstd{);}
\end{alltt}


{\ttfamily\noindent\color{warningcolor}{Warning in model.response(mf, "{}numeric"{}): using type = "{}numeric"{} with a factor response will be ignored}}

{\ttfamily\noindent\color{warningcolor}{Warning in Ops.factor(y, z\$residuals): '-' not meaningful for factors}}\end{kframe}

{\centering \includegraphics[width=.49\textwidth]{graphics/plotunnamed-chunk-8-1} 

}



\end{knitrout}

Plot in 3d. \verb|rgl| provides an interactive 3d box.
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlkwd{require}\hlstd{(scatterplot3d)}
\hlstd{scat} \hlkwb{<-} \hlkwd{with}\hlstd{(top100,}
             \hlkwd{scatterplot3d}\hlstd{(Times.cited,}
                           \hlstd{Year, Rank,} \hlkwc{pch}\hlstd{=}\hlnum{16}\hlstd{,}
                           \hlkwc{highlight.3d}\hlstd{=}\hlnum{TRUE}\hlstd{,}
                           \hlkwc{type}\hlstd{=}\hlstr{"h"}
                           \hlstd{)}
             \hlstd{)}
\hlstd{fit} \hlkwb{<-} \hlkwd{with}\hlstd{(top100,} \hlkwd{lm}\hlstd{(Rank}\hlopt{~}\hlstd{Times.cited}\hlopt{+}\hlstd{Year))}
\end{alltt}


{\ttfamily\noindent\color{warningcolor}{Warning in model.response(mf, "{}numeric"{}): using type = "{}numeric"{} with a factor response will be ignored}}

{\ttfamily\noindent\color{warningcolor}{Warning in Ops.factor(y, z\$residuals): '-' not meaningful for factors}}\begin{alltt}
\hlstd{scat}\hlopt{$}\hlkwd{plane3d}\hlstd{(fit)}
\end{alltt}
\end{kframe}

{\centering \includegraphics[width=.49\textwidth]{graphics/plotunnamed-chunk-9-1} 

}



\end{knitrout}



\newpage
\section{System Information}
\label{sec:sys_info}

The version number of R and packages loaded for generating the vignette were:
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{verbatim}
R version 3.1.2 (2014-10-31)
Platform: x86_64-unknown-linux-gnu (64-bit)

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
 [1] splines   grid      parallel  stats     graphics  grDevices utils     datasets 
 [9] methods   base     

other attached packages:
 [1] scatterplot3d_0.3-35 vegan_2.2-0          permute_0.8-3       
 [4] reshape_0.8.5        httr_0.5             rgl_0.95.1158       
 [7] plyr_1.8.1           Hmisc_3.14-6         Formula_1.1-2       
[10] survival_2.37-7      lattice_0.20-29      doParallel_1.0.8    
[13] iterators_1.0.7      foreach_1.4.2        rbenchmark_1.0.0    
[16] ggplot2_1.0.0        dplyr_0.3.0.2        knitr_1.8           

loaded via a namespace (and not attached):
 [1] acepack_1.3-3.3     assertthat_0.1      cluster_1.15.3      codetools_0.2-9    
 [5] colorspace_1.2-4    compiler_3.1.2      DBI_0.3.1           digest_0.6.4       
 [9] evaluate_0.5.5      foreign_0.8-61      formatR_1.0         gtable_0.1.2       
[13] highr_0.4           labeling_0.3        latticeExtra_0.6-26 lazyeval_0.1.9     
[17] magrittr_1.5        MASS_7.3-35         Matrix_1.1-4        mgcv_1.8-3         
[21] munsell_0.4.2       nlme_3.1-118        nnet_7.3-8          proto_0.3-10       
[25] RColorBrewer_1.0-5  Rcpp_0.11.3         reshape2_1.4        rpart_4.1-8        
[29] scales_0.2.4        stringr_0.6.2       tcltk_3.1.2         tools_3.1.2        
\end{verbatim}
\end{kframe}
\end{knitrout}


\end{linenumbers}
\end{document}



